File: APIs/app/controllers/newsSubscriptionController/subscribe_daily_mail_controller.py
from fastapi import Response, HTTPException, Request
from pydantic import BaseModel, EmailStr
from app.database.models.subscription_model import SubscribeMail
from app.utils.email_template import subscribe_daily_mail_email_data, send_daily_tip_email_data
from app.services.mail_service import send_mail
from app.utils.gemini_utils.generate_daily_tips_prompt import generate_daily_tips_prompt
from app.services.groq_service import groq_generated_data
from app.services.winston_service import get_logger
import json

logger = get_logger()

class SubscribeRequest(BaseModel):
    userMail: EmailStr

async def subscribe_daily_mail_controller(request: Request, body: SubscribeRequest, response: Response):
    full_url = str(request.url)
    try:
        user_mail = body.userMail
        
        # Check exist
        exist_mail = await SubscribeMail.find_one({"mail": user_mail})
        
        if exist_mail:
            logger.info(f"URL: {full_url} - Already subscribed")
            return {"status": "OK", "message": "Already subscribed!"}
            
        # Create
        new_sub = SubscribeMail(mail=user_mail)
        await new_sub.save()
        
        # 1. Send Welcome Mail
        mail_data = subscribe_daily_mail_email_data(user_mail)
        await send_mail(mail_data['to'], mail_data['subject'], mail_data['html'])
        
        # 2. Generate Daily Tips
        prompt = generate_daily_tips_prompt()
        daily_tips_content = await groq_generated_data(prompt)
        
        try:
            start_index = daily_tips_content.index('{')
            end_index = daily_tips_content.rindex('}')
            clean_string = daily_tips_content[start_index:end_index+1]
            tip_data_object = json.loads(clean_string)
            
            # 3. Send Tip Mail
            tip_mail_data = send_daily_tip_email_data(user_mail, tip_data_object)
            await send_mail(tip_mail_data['to'], tip_mail_data['subject'], tip_mail_data['html'])
            
        except Exception as e:
            logger.error(f"Error parsing/sending daily tip: {e}")
            # Continue execution as user is registered
            
        logger.info(f"URL: {full_url} - Registered")
        return {"status": "Ok", "message": "Registered!"}
        
    except Exception as e:
        logger.error(f"URL: {full_url}, error: {e}")
        raise HTTPException(status_code=500, detail="Internal Server Error")

File: Backend/src/controllers/newsSubscriptionController/subscribeDailyMail.controller.ts
import { getReasonPhrase, StatusCodes } from "http-status-codes"
import SubscribeMail from "../../database/models/subscribeMail.model";
import emailTemplates from "../../utils/email-templates";
import sendMail from "../../services/mailService";
import { generateDailyTips } from "../../utils/gemini/generateDailyTips.prompt";

import { groqGeneratedData } from "../../services/groq.service";
import logger from "../../utils/logger";
import getFullURL from "../../services/getFullURL.service";

/**
 * Controller to handle Daily Mail Subscriptions.
 * Verifies email, creates subscription, sends welcome mail, generates first tip, and sends it.
 */
const subscribeDailyMailController = async (req, res) => {
    const fullUrl: string = getFullURL(req);
    try {
        const { userMail } = req.body;

        // 1. Verify Request: Ensure email is provided
        if (!userMail) {
            logger.error(`URL: ${fullUrl} - User mail is not provided.`);
            return res.status(StatusCodes.BAD_REQUEST).json({
                status: "Failed",
                message: "Required fields not exist!"
            });
        }

        // 2. Check for Existing Subscription
        const existMail = await SubscribeMail.findOne({ mail: userMail });

        if (existMail) {
            logger.info(`URL: ${fullUrl} - Already subscribed`);
            return res.status(StatusCodes.OK).json({
                status: 'OK',
                message: "Already subscribed!"
            });
        }

        // 3. Create New Subscription in Database
        // Note: Field name mismatch? 'mail' vs 'userMail' in schema? Assuming 'mail' based on existing code, but schema has 'userMail'.
        // FIXME: Check model schema match.
        const newSubscribeMail = new SubscribeMail({
            mail: userMail
        });

        await newSubscribeMail.save();

        // 4. Send Welcome Email
        let mailData = emailTemplates.subscribeDailyMailEmailData(userMail);

        await sendMail(mailData, (mailError: Error | null) => {
            logger.info(`URL: ${fullUrl}`);
            if (mailError) {
                logger.error(`URL: ${fullUrl} - Failed to send welcome mail. Error: ${mailError.message}`);
                // Note: We don't return here so execution continues to generate tip?
                // Or should we fail?
                return res.status(StatusCodes.EXPECTATION_FAILED).json({
                    status: 'Failed',
                    message: "Mail sending error!"
                });
            }
        });

        // 5. Generate First Daily Tip using AI
        const prompt = generateDailyTips();
        const generateDailyTipsContent = await groqGeneratedData(prompt);

        // 6. Clean AI Response (Extract JSON)
        const startIndex = generateDailyTipsContent.indexOf('{');
        const endIndex = generateDailyTipsContent.lastIndexOf('}');
        const cleanString = generateDailyTipsContent.substring(startIndex, endIndex + 1);

        const tipDataObject = JSON.parse(cleanString);

        // 7. Send Daily Tip Email
        mailData = emailTemplates.sendDailyTipEmailData(userMail, tipDataObject);

        await sendMail(mailData, (mailError: Error | null) => {
            logger.info(`URL: ${fullUrl}`);
            if (mailError) {
                logger.error(`URL: ${fullUrl} - Failed to send first travel tips mail. Error: ${mailError.message}`);
                return res.status(StatusCodes.EXPECTATION_FAILED).json({
                    status: 'Failed',
                    message: "Mail sending error!"
                });
            }
        });

        // 8. Success Response
        logger.info(`URL: ${fullUrl} - Registered`);
        return res.status(StatusCodes.OK).json({
            status: "Ok",
            message: "Registered!",
        });
    } catch (error) {
        logger.error(`URL: ${fullUrl}, error_message: ${error.message}`);
        return res.status(StatusCodes.INTERNAL_SERVER_ERROR);
    }
}

export default subscribeDailyMailController;